Build 34 for 8.3 with status FAILURE

## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2F8.3/detail/8.3/34//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2F8.3/detail/8.3/34//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2F8.3/detail/8.3/34//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2F8.3/detail/8.3/34//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_null.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-06-23T04:38:27.855Z&rangeTo=2022-06-23T04:58:27.855Z&transactionName=Beats/beats/8.3&transactionType=job&latencyAggregationType=avg&traceId=108fe044a735bd4904609294be28c308&transactionId=962be26e6eb473e5)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-06-23T04:48:27.855+0000


* Duration: 141 min 33 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 44523  |
| Skipped      | 4950 |
| Total        | 49473   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2F8.3/detail/8.3/34//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `filebeat-windows-8-windows-8 - mage build unitTest`
<ul>
<li>Took 6 min 58 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/8.3/runs/34/steps/32242/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `filebeat-windows-8-windows-8 - mage build unitTest`
<ul>
<li>Took 4 min 50 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/8.3/runs/34/steps/34433/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `filebeat-windows-8-windows-8 - mage build unitTest`
<ul>
<li>Took 4 min 49 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/8.3/runs/34/steps/36015/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/8.3/runs/34/steps/36944/log/?start=0">here</a></li>
<li>Description: <code>Error "hudson.AbortException: script returned exit code 1"</code></l1>
</ul>
  
  </p>
  </details>

```
2022-06-23T06:58:52.627Z] ================================== FAILURES ===================================
[2022-06-23T06:58:52.627Z] _________________ TestKeystore.test_keystore_with_present_key _________________
[2022-06-23T06:58:52.627Z] 
[2022-06-23T06:58:52.627Z] self = <test_keystore.TestKeystore testMethod=test_keystore_with_present_key>
[2022-06-23T06:58:52.627Z] 
[2022-06-23T06:58:52.627Z]     def setUp(self):
[2022-06-23T06:58:52.627Z] >       super(BaseTest, self).setUp()
[2022-06-23T06:58:52.627Z] 
[2022-06-23T06:58:52.627Z] tests\system\test_keystore.py:14: 
[2022-06-23T06:58:52.627Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2022-06-23T06:58:52.627Z] ..\libbeat\tests\system\beat\beat.py:401: in setUp
[2022-06-23T06:58:52.627Z]     shutil.rmtree(self.working_dir)
[2022-06-23T06:58:52.627Z] C:\Python38\lib\shutil.py:737: in rmtree
[2022-06-23T06:58:52.627Z]     return _rmtree_unsafe(path, onerror)
[2022-06-23T06:58:52.627Z] C:\Python38\lib\shutil.py:610: in _rmtree_unsafe
[2022-06-23T06:58:52.627Z]     _rmtree_unsafe(fullname, onerror)
[2022-06-23T06:58:52.627Z] C:\Python38\lib\shutil.py:615: in _rmtree_unsafe
[2022-06-23T06:58:52.627Z]     onerror(os.unlink, fullname, sys.exc_info())
[2022-06-23T06:58:52.627Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2022-06-23T06:58:52.627Z] 
[2022-06-23T06:58:52.627Z] path = 'C:\\Users\\jenkins\\workspace\\8.3-34-6fa18d96-b1b3-45cc-b5a2-32fedde40cfa\\src\\github.com\\elastic\\beats\\filebeat\\build\\system-tests\\run\\test_keystore.TestKeystore.test_keystore_with_present_key\\data'
[2022-06-23T06:58:52.628Z] onerror = <function rmtree.<locals>.onerror at 0x000000040F6F03A0>
[2022-06-23T06:58:52.628Z] 
[2022-06-23T06:58:52.628Z]     def _rmtree_unsafe(path, onerror):
[2022-06-23T06:58:52.628Z]         try:
[2022-06-23T06:58:52.628Z]             with os.scandir(path) as scandir_it:
[2022-06-23T06:58:52.628Z]                 entries = list(scandir_it)
[2022-06-23T06:58:52.628Z]         except OSError:
[2022-06-23T06:58:52.628Z]             onerror(os.scandir, path, sys.exc_info())
[2022-06-23T06:58:52.628Z]             entries = []
[2022-06-23T06:58:52.628Z]         for entry in entries:
[2022-06-23T06:58:52.628Z]             fullname = entry.path
[2022-06-23T06:58:52.628Z]             if _rmtree_isdir(entry):
[2022-06-23T06:58:52.628Z]                 try:
[2022-06-23T06:58:52.628Z]                     if entry.is_symlink():
[2022-06-23T06:58:52.628Z]                         # This can only happen if someone replaces
[2022-06-23T06:58:52.628Z]                         # a directory with a symlink after the call to
[2022-06-23T06:58:52.628Z]                         # os.scandir or entry.is_dir above.
[2022-06-23T06:58:52.628Z]                         raise OSError("Cannot call rmtree on a symbolic link")
[2022-06-23T06:58:52.628Z]                 except OSError:
[2022-06-23T06:58:52.628Z]                     onerror(os.path.islink, fullname, sys.exc_info())
[2022-06-23T06:58:52.628Z]                     continue
[2022-06-23T06:58:52.628Z]                 _rmtree_unsafe(fullname, onerror)
[2022-06-23T06:58:52.628Z]             else:
[2022-06-23T06:58:52.628Z]                 try:
[2022-06-23T06:58:52.628Z] >                   os.unlink(fullname)
[2022-06-23T06:58:52.628Z] E                   PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\jenkins\\workspace\\8.3-34-6fa18d96-b1b3-45cc-b5a2-32fedde40cfa\\src\\github.com\\elastic\\beats\\filebeat\\build\\system-tests\\run\\test_keystore.TestKeystore.test_keystore_with_present_key\\data\\filebeat.lock'
[2022-06-23T06:58:52.628Z] 
[2022-06-23T06:58:52.628Z] C:\Python38\lib\shutil.py:613: PermissionError
```

--- a/libbeat/tests/system/beat/beat.py
+++ b/libbeat/tests/system/beat/beat.py
@@ -3,7 +3,7 @@ beat defines the basic testing infrastructure used by individual python unit/int
 """
 
 import subprocess
-
+import random
 import unittest
 import os
 import shutil
@@ -396,7 +396,7 @@ class TestCase(unittest.TestCase, ComposeMixin):
 
         # create working dir
         self.working_dir = os.path.abspath(os.path.join(
-            self.build_path + "run", self.id()))
+            self.build_path, "run", self.id() + str(random.randrange(1000))))
         if os.path.exists(self.working_dir):
             shutil.rmtree(self.working_dir)
         os.makedirs(self.working_dir)
@@ -411,12 +411,24 @@ class TestCase(unittest.TestCase, ComposeMixin):
             # update the last_run link
             if os.path.islink(self.build_path + "last_run"):
                 os.unlink(self.build_path + "last_run")
-            os.symlink(self.build_path + f"run/{self.id()}",
+            os.symlink(self.working_dir,
                        self.build_path + "last_run")
         except BaseException:
             # symlink is best effort and can fail when
             # running tests in parallel
             pass
+
+        # Keep last 5 runs
+        candidates = []
+        to_keep = 5
+        for dir_entry in os.listdir(os.path.join(self.build_path, "run")):
+            if re.search(self.id() + r"[0-9]+$", dir_entry):
+                candidates.append(dir_entry)
+        if len(candidates) > to_keep:
+            candidates.sort(reverse=True, key=lambda dirname: os.path.getmtime(
+                os.path.join(self.build_path, "run", dirname)))
+            for d in candidates[to_keep:]:
+                shutil.rmtree(os.path.join(self.build_path, "run", d))
 
     def wait_until(self, cond, max_timeout=20, poll_interval=0.1, name="cond", err_msg=""):
         """

